### PR TITLE
Create regulariser object and MCMC particle regulariser

### PR DESCRIPTION
**Pull request by timothy-glover**
Created the regulariser object for optional use with particle filters when impoverishment is a problem. This has included the class MCMCRegulariser that follows a particular implementation using the Metropolis-Hastings algorithm (referenced appropriately in the docs).

### CODE REVIEW

File: stonesoup/regulariser/particle.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            part_diff = np.array(moved_particles.state_vector - prior.state_vector).T
+            part_diff_mean = np.mean(part_diff, axis=0)
+            move_likelihood = multivariate_normal.pdf(part_diff,
+                                                      mean=part_diff_mean,
+                                                      cov=np.array(covar_est))
+            post_part_diff = np.array(posterior.state_vector - prior.state_vector).T
+            post_part_diff_mean = np.mean(post_part_diff, axis=0)
+            post_likelihood = multivariate_normal.pdf(post_part_diff,
+                                                      mean=post_part_diff_mean,
+                                                      cov=covar_est)
```
Comment (sdhiscocks) [inline stonesoup/regulariser/particle.py:84]: The underlying `pdf` method does `x - mean` after converting everything to standard float array, but by calculating difference prior to calling `pdf` and leaving the mean as `0`, this allows the custom typing in `StateVectors` to be used e.g. handle angle wrapping.

I also swapped `mean` to `average` as it turns out `mean` override on `StateVectors` is completely broken at the moment, which I'll fix in another PR.

```suggestion
            # Evaluate likelihoods
            part_diff = moved_particles.state_vector - prior.state_vector
            part_diff_mean = np.average(part_diff, axis=1)
            move_likelihood = multivariate_normal.pdf((part_diff - part_diff_mean).T,
                                                      cov=covar_est)
            post_part_diff = posterior.state_vector - prior.state_vector
            post_part_diff_mean = np.average(post_part_diff, axis=1)
            post_likelihood = multivariate_normal.pdf((post_part_diff - post_part_diff_mean).T,
                                                      cov=covar_est)
```

File: stonesoup/regulariser/particle.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+            # Evaluate measurement likelihoods
+            move_meas_likelihood = []
+            post_meas_likelihood = []
+            for detection in detections:
+                move_meas_likelihood.append(measurement_model.logpdf(detection, moved_particles))
+                post_meas_likelihood.append(measurement_model.logpdf(detection, moved_particles))
```
Comment (sdhiscocks) [inline stonesoup/regulariser/particle.py:91]: Just to save calling `logpdf` twice.
```suggestion
            for detection in detections:
                likelihood = measurement_model.logpdf(detection, moved_particles)
                move_meas_likelihood.append(likelihood)
                post_meas_likelihood.append(likelihood)
```

File: stonesoup/regulariser/particle.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+            # In the case that there are multiple measurements,
+            # we select the highest overall likelihood.
+            max_likelihood_idx = np.argmax(np.sum(move_meas_likelihood, axis=1))
+
+            # Calculate acceptance probability (alpha)
+            alpha = (move_meas_likelihood[max_likelihood_idx]*move_likelihood) / \
+                    (post_meas_likelihood[max_likelihood_idx]*post_likelihood)
```
Comment (sdhiscocks) [inline stonesoup/regulariser/particle.py:99]: `move_likelihood` and `post_likelihood` are from `pdf`, and `move_meas_likelihood` and `post_meas_likelihood` are using `logpdf`. Is this right?

File: stonesoup/regulariser/particle.py (hunk 2 above)
Comment (sdhiscocks) [reply stonesoup/regulariser/particle.py:91]: Actually, is this an error, and they shouldn't be the same? From variable names looks like it should be `posterior`.
```suggestion
            for detection in detections:
                move_meas_likelihood.append(measurement_model.logpdf(detection, moved_particles))
                post_meas_likelihood.append(measurement_model.logpdf(detection, posterior))
```

**Review by sdhiscocks — CHANGES_REQUESTED** [review]

File: stonesoup/regulariser/particle.py (hunk 3 above)
Comment (timothy-glover) [reply stonesoup/regulariser/particle.py]: You are right, there is a mismatch between `logpdf` and `pdf`. I have changed the `move_likelihood` and `post_likelihood` to also be `logpdf` and changed the calculation of `alpha` accordingly.

**Review by sdhiscocks — APPROVED** [review]

**Review by jswright-dstl — APPROVED** [review]

**Pull request merged**